File: lib/models/user_model.dart
// Modelo de dados do usuário utilizado no app Flutter
class User {
  final String? id; // ID do usuário (MongoDB)
  final String nome; // Nome completo
  final String email; // Email
  final String matricula; // Matrícula institucional
  final String role; // Papel: aluno, professor ou admin
  final String? curso; // Curso (apenas para alunos)
  final List<String> turmas; // Lista de turmas
  final int saldo; // Saldo de coins
  final String? fotoPerfil; // URL ou caminho da foto de perfil
  final String? statusAprovacao; // Status de aprovação (professor)
  final DateTime? createdAt; // Data de criação
  final DateTime? updatedAt; // Data de atualização

  User({
    this.id,
    required this.nome,
    required this.email,
    required this.matricula,
    required this.role,
    this.curso,
    required this.turmas,
    required this.saldo,
    this.fotoPerfil,
    this.statusAprovacao,
    this.createdAt,
    this.updatedAt,
  });

  // Cria um User a partir de um JSON
  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      id: json['_id']?.toString() ?? '',
      nome: (json['nome'] ?? '').toString(),
      email: (json['email'] ?? '').toString(),
      matricula: (json['matricula'] ?? '').toString(),
      role: (json['role'] ?? '').toString(),
      curso: json['curso']?.toString(),
      turmas: json['turmas'] != null
          ? List<String>.from((json['turmas'] as List).map((e) => e?.toString() ?? ''))
          : <String>[],
      saldo: json['saldo'] is int
          ? json['saldo']
          : int.tryParse(json['saldo']?.toString() ?? '0') ?? 0,
      fotoPerfil: json['fotoPerfil']?.toString(),
      statusAprovacao: json['statusAprovacao']?.toString(),
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString())
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'].toString())
          : null,
    );
  }

  // Converte o User para JSON
  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'nome': nome,
      'email': email,
      'matricula': matricula,
      'role': role,
      'curso': curso,
      'turmas': turmas,
      'saldo': saldo,
      'fotoPerfil': fotoPerfil,
      'statusAprovacao': statusAprovacao,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  // Cria uma cópia do usuário com campos modificados
  User copyWith({
    String? id,
    String? nome,
    String? email,
    String? matricula,
    String? role,
    String? curso,
    List<String>? turmas,
    int? saldo,
    String? fotoPerfil,
    String? statusAprovacao,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return User(
      id: id ?? this.id,
      nome: nome ?? this.nome,
      email: email ?? this.email,
      matricula: matricula ?? this.matricula,
      role: role ?? this.role,
      curso: curso ?? this.curso,
      turmas: turmas ?? this.turmas,
      saldo: saldo ?? this.saldo,
      fotoPerfil: fotoPerfil ?? this.fotoPerfil,
      statusAprovacao: statusAprovacao ?? this.statusAprovacao,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  bool get isAdminOrProfessor => role == 'admin' || role == 'professor';
}

// Classe auxiliar para requisição de login
class LoginRequest {
  final String matricula;
  final String senha;

  LoginRequest({
    required this.matricula,
    required this.senha,
  });

  Map<String, dynamic> toJson() {
    return {
      'matricula': matricula,
      'senha': senha,
    };
  }
}

// Classe auxiliar para resposta de login
class LoginResponse {
  final String token;
  final User user;
  final bool showApprovalNotification;

  LoginResponse({
    required this.token,
    required this.user,
    this.showApprovalNotification = false,
  });

  factory LoginResponse.fromJson(Map<String, dynamic> json) {
    return LoginResponse(
      token: json['token'],
      user: User.fromJson(json['user']),
      showApprovalNotification: json['showApprovalNotification'] ?? false,
    );
  }
} 

File: backend/models/userModel.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');

// Schema do usuário, define a estrutura dos documentos na coleção 'users'
const userSchema = new mongoose.Schema({
    nome: {
        type: String,
        required: [true, 'Nome é obrigatório'],
        trim: true
    },
    email: { 
        type: String, 
        unique: true,
        required: [true, 'Email é obrigatório'],
        lowercase: true,
        trim: true,
        match: [/^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/, 'Email inválido'] // Regex para validar email
    },
    senha: {
        type: String,
        required: [true, 'Senha é obrigatória'],
        minlength: [6, 'Senha deve ter pelo menos 6 caracteres']
    },
    matricula: { 
        type: String, 
        unique: true,
        required: [true, 'Matrícula é obrigatória'],
        trim: true
    },
    role: { 
        type: String, 
        enum: ['aluno', 'professor', 'admin'], 
        default: 'aluno' // Define o tipo de usuário
    },
    statusAprovacao: {
        type: String,
        enum: ['pendente', 'aprovado', 'recusado'],
        default: function() {
            // Professores começam como pendentes, outros como aprovados
            return this.role === 'professor' ? 'pendente' : 'aprovado';
        }
    },
    curso: { 
        type: String, 
        enum: ['Engenharia de Alimentos', 'Agropecuária', 'Informática para Internet'],
        required: function() {
            // Curso é obrigatório apenas para alunos
            return this.role === 'aluno';
        }
    },
    turmas: [String], // Lista de turmas do usuário
    saldo: { 
        type: Number, 
        default: 0,
        min: [0, 'Saldo não pode ser negativo']
    },
    fotoPerfil: { 
        type: String, 
        default: '' // URL ou caminho da foto de perfil
    },
    fotoPerfilBin: {
        type: Buffer,
        select: false //não retorna por padrão
    },
    ultimoLogin: {
        type: Date,
        default: Date.now //daata do último login
    },
    ativo: {
        type: Boolean,
        default: true //usuário ativo ou não
    }
}, 
{ timestamps: true }); //adiciona createdAt e updatedAt automaticamente

// Middleware para hash da senha antes de salvar o usuário
userSchema.pre('save', async function(next) {
    // Só faz hash se a senha foi modificada ou é novo usuário
    if (!this.isModified('senha')) return next();
    
    try {
        // Gera salt e faz hash da senha
        const salt = await bcrypt.genSalt(12);
        this.senha = await bcrypt.hash(this.senha, salt);
        next();
    } catch (error) {
        next(error);
    }
});

// Método de instância para comparar senha informada com a salva (hash)
userSchema.methods.compararSenha = async function(senhaCandidata) {
    return await bcrypt.compare(senhaCandidata, this.senha);
};

// Método para adicionar coins ao saldo do usuário
userSchema.methods.adicionarCoins = function(quantidade) {
    if (quantidade > 0) {
        this.saldo += quantidade;
        return this.save();
    }
    throw new Error('Quantidade deve ser positiva');
};

// Método para remover coins do saldo do usuário
userSchema.methods.removerCoins = function(quantidade) {
    if (quantidade > 0 && this.saldo >= quantidade) {
        this.saldo -= quantidade;
        return this.save();
    }
    throw new Error('Saldo insuficiente ou quantidade inválida');
};

// Método para atualizar a data do último login
userSchema.methods.atualizarUltimoLogin = function() {
    this.ultimoLogin = new Date();
    return this.save();
};

// Método para retornar os dados públicos do usuário (sem senha e foto binária)
userSchema.methods.toPublicJSON = function() {
    const userObject = this.toObject();
    delete userObject.senha;
    delete userObject.fotoPerfilBin;
    return userObject;
};

// Índices para melhorar performance em buscas por role e ativo
userSchema.index({ role: 1 });
userSchema.index({ ativo: 1 });

// Exporta o modelo User para ser usado em outras partes do projeto
module.exports = mongoose.model('User', userSchema);
